{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cff228f6-fac0-4ba7-949e-6fa04af542b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read in gene expression matrices for X and Y and the two files \n",
    "#containing information on the differentially expressed genes\n",
    "\n",
    "GEM_X_file = open('GEM_X.csv', 'r')\n",
    "GEM_Y_file = open('GEM_Y.csv', 'r')\n",
    "variety_X_file = open('Leaf_DEGs_VarX_T1.csv', 'r')\n",
    "variety_Y_file = open('Leaf_DEGs_VarY_T1.csv', 'r')\n",
    "\n",
    "#empty lists to hold split lines of csv files\n",
    "GEMX_split = []\n",
    "GEMY_split = []\n",
    "VARY_split = []\n",
    "VARX_split = []\n",
    "\n",
    "#read lines of GEM X file\n",
    "for line in GEM_X_file.readlines():\n",
    "    line_split = line.split(',')\n",
    "    GEMX_split.append(line_split)\n",
    "\n",
    "#read lines of GEM Y file\n",
    "for line in GEM_Y_file.readlines():\n",
    "    line_split = line.split(',')\n",
    "    GEMY_split.append(line_split)\n",
    "\n",
    "#read lines of Variety X file\n",
    "for line in variety_X_file.readlines():\n",
    "    line_split = line.split(',')\n",
    "    VARX_split.append(line_split)\n",
    "\n",
    "#read lines of Variety Y file\n",
    "for line in variety_Y_file.readlines():\n",
    "    line_split = line.split(',')\n",
    "    VARY_split.append(line_split)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e3ae24fd-7c6a-455f-af51-0a0ce8283541",
   "metadata": {},
   "outputs": [],
   "source": [
    "#strip all values of trailing white spaces - to avoid errors in R - ie things not matching\n",
    "\n",
    "#rid of trailing white spaces for GEMX\n",
    "for i in range(len(GEMX_split)):\n",
    "    for i2 in range(len(GEMX_split[i])):\n",
    "        GEMX_split[i][i2] == GEMX_split[i][i2].rstrip()\n",
    "        \n",
    "#rid of trailing white spaces for GEMY\n",
    "for i in range(len(GEMY_split)):\n",
    "    for i2 in range(len(GEMY_split[i])):\n",
    "        GEMY_split[i][i2] == GEMY_split[i][i2].rstrip()\n",
    "\n",
    "#rid of trailing white spaces for VARX\n",
    "for i in range(len(VARX_split)):\n",
    "    for i2 in range(len(VARX_split[i])):\n",
    "        VARX_split[i][i2] == VARX_split[i][i2].rstrip()\n",
    "        \n",
    "#rid of trailing white spaces for VARY\n",
    "for i in range(len(VARY_split)):\n",
    "    for i2 in range(len(VARY_split[i])):\n",
    "        VARY_split[i][i2] == VARY_split[i][i2].rstrip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5253469f-6d89-4d59-a536-91d1baf3d592",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create ouput: Two gene expression matrices (one for each Variety), \n",
    "#each with a header, and containing gene expression from two conditions only: \n",
    "#the control condition (in columns 2-4; replicates 1-3) \n",
    "#and stress treatment condition 1 (columns 5-7; replicates 1-3). \n",
    "\n",
    "#empty lists to hold reformated gene expression matrices\n",
    "gemX = []\n",
    "gemY = []\n",
    "leafX = []\n",
    "leafY = []\n",
    "\n",
    "#first append the headers to each list in the order wanted\n",
    "gemXheaders = [GEMX_split[0][0],GEMX_split[0][1],GEMX_split[0][4],GEMX_split[0][8],GEMX_split[0][2],GEMX_split[0][6],GEMX_split[0][10]+'\\n']\n",
    "gemX.append(gemXheaders)\n",
    "\n",
    "gemYheaders = [GEMY_split[0][0],GEMY_split[0][1],GEMY_split[0][5],GEMY_split[0][9],GEMY_split[0][3],GEMY_split[0][7],GEMY_split[0][11]]\n",
    "gemY.append(gemYheaders)\n",
    "\n",
    "leafXheaders = ['gene_name',VARX_split[0][2],VARX_split[0][6][:-1],VARX_split[0][4],VARX_split[0][5],GEMX_split[0][1],GEMX_split[0][4],GEMX_split[0][8],GEMX_split[0][2],GEMX_split[0][6],GEMX_split[0][10]+'\\n']\n",
    "leafX.append(leafXheaders)\n",
    "\n",
    "leafYheaders = ['gene_name',VARY_split[0][2],VARY_split[0][6][:-1],VARY_split[0][4],VARY_split[0][5],GEMY_split[0][1],GEMY_split[0][5],GEMY_split[0][9],GEMY_split[0][3],GEMY_split[0][7],GEMY_split[0][11]]\n",
    "leafY.append(leafYheaders)\n",
    "\n",
    "#sort data into newly organized rows to append to each of gemX, gemY, leafX, leafY \n",
    "for i in range(1,len(GEMX_split)):\n",
    "    newX_row = [GEMX_split[i][0],GEMX_split[i][1],GEMX_split[i][4],GEMX_split[i][8],GEMX_split[i][2],GEMX_split[i][6],GEMX_split[i][10]+'\\n']\n",
    "    gemX.append(newX_row)\n",
    "    \n",
    "for i in range(1,len(GEMY_split)):\n",
    "    newY_row = [GEMY_split[i][0],GEMY_split[i][1],GEMY_split[i][5],GEMY_split[i][9],GEMY_split[i][3],GEMY_split[i][7],GEMY_split[i][11]]\n",
    "    gemY.append(newY_row)\n",
    "    \n",
    "#intialize index variable\n",
    "index = -1\n",
    "\n",
    "#for each gene in VARX, iterating over gemX to find the matching gene\n",
    "for i in range(1, len(VARX_split)):\n",
    "    \n",
    "    gene_to_find = VARX_split[i][0]\n",
    "    \n",
    "    for x, gene in enumerate(gemX):\n",
    "        if gene[0] == gene_to_find:\n",
    "            index = x\n",
    "    \n",
    "    #once matching index is found, can access replicate numbers and append new row to leafX\n",
    "    leafX_row = [gene_to_find,VARX_split[i][2],VARX_split[i][6],VARX_split[i][4],VARX_split[i][5],gemX[index][1],gemX[index][2],gemX[index][3],gemX[index][3],gemX[index][5],gemX[index][6]]\n",
    "    leafX.append(leafX_row)\n",
    "\n",
    "#Same steps but for Leaf Variety Y\n",
    "#intialize index variable\n",
    "indY = -1\n",
    "\n",
    "#for each gene in VARY, iterating over gemY to find the matching gene\n",
    "for i in range(1, len(VARY_split)):\n",
    "    \n",
    "    gene_to_find = VARY_split[i][0]\n",
    "    \n",
    "    for y, geneY in enumerate(gemY):\n",
    "        if geneY[0] == gene_to_find:\n",
    "            indY = y\n",
    "    \n",
    "    #once matching index is found, can access replicate numbers and append new row to leafY\n",
    "    leafY_row = [gene_to_find,VARY_split[i][2],VARY_split[i][6],VARY_split[i][4],VARY_split[i][5],gemY[indY][1],gemY[indY][2],gemY[indY][3],gemY[indY][3],gemY[indY][5],gemY[indY][6]]\n",
    "    leafY.append(leafY_row)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b5a6613-a4b2-48f6-9c0a-6b2fa5c181cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create csv file outputs for each of the outputs above\n",
    "\n",
    "import csv\n",
    "\n",
    "#list gemX and gemY are the two gene expression matrices containing gene expression for control and stress conditions\n",
    "\n",
    "#gemX csv will be all_VarX_TwoTimePoints.csv\n",
    "with open('all_VarX_TwoTimePoints.csv', mode='w', newline='') as csvfile:\n",
    "    #create file to write with csv writer\n",
    "    csv_wr = csv.writer(csvfile)\n",
    "    #write each row from data_reform to csv file\n",
    "    csv_wr.writerows(gemX)\n",
    "    \n",
    "#gemY csv will be all_VarY_TwoTimePoints.csv\n",
    "with open('all_VarY_TwoTimePoints.csv', mode='w', newline='') as csvfile:\n",
    "    #create file to write with csv writer\n",
    "    csv_wr = csv.writer(csvfile)\n",
    "    #write each row from data_reform to csv file\n",
    "    csv_wr.writerows(gemY)\n",
    "    \n",
    "#leafX csv will be Leaf_DEGs_VarX.csv\n",
    "with open('Leaf_DEGs_VarX.csv', mode='w', newline='') as csvfile:\n",
    "    #create file to write with csv writer\n",
    "    csv_wr = csv.writer(csvfile)\n",
    "    #write each row from data_reform to csv file\n",
    "    csv_wr.writerows(leafX)\n",
    "    \n",
    "#leafY csv will be Leaf_DEGs_VarY.csv\n",
    "with open('Leaf_DEGs_VarY.csv', mode='w', newline='') as csvfile:\n",
    "    #create file to write with csv writer\n",
    "    csv_wr = csv.writer(csvfile)\n",
    "    #write each row from data_reform to csv file\n",
    "    csv_wr.writerows(leafY)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
